### PR TITLE
350 pairing extra feats

### PR DESCRIPTION
**Pull request by cpaniaguam**
#350 #339

### CODE REVIEW

**Comment by cpaniaguam** [comment]
@danielmwatkins Below is what the output data structure is looking like after the matching criteria is addee. All units for lengths and areas are in km and sq. km, respectively; orientation is in radians. `area_under` (as named in the Matlab script) is the registration "floe mismatch" (sometimes clamped to 0 if a certain threshold is met);  `corr` is the psi-s correlation. They are computed between the floes in the current row and the next. Note that these values are `missing` for the last floe in each group `ID`. 
<div><div style = "float: left;"><span>6×14 DataFrame</span></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "header"><th class = "rowNumber" style = "font-weight: bold; text-align: right;">Row</th><th style = "text-align: left;">ID</th><th style = "text-align: left;">passtime</th><th style = "text-align: left;">area</th><th style = "text-align: left;">convex_area</th><th style = "text-align: left;">major_axis_length</th><th style = "text-align: left;">minor_axis_length</th><th style = "text-align: left;">orientation</th><th style = "text-align: left;">perimeter</th><th style = "text-align: left;">latitude</th><th style = "text-align: left;">longitude</th><th style = "text-align: left;">x</th><th style = "text-align: left;">y</th><th style = "text-align: left;">area_under</th><th style = "text-align: left;">corr</th></tr><tr class = "subheader headerLastRow"><th class = "rowNumber" style = "font-weight: bold; text-align: right;"></th><th title = "Int64" style = "text-align: left;">Int64</th><th title = "Dates.DateTime" style = "text-align: left;">DateTime</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Union{Missing, Float64}" style = "text-align: left;">Float64?</th><th title = "Union{Missing, Float64}" style = "text-align: left;">Float64?</th></tr></thead><tbody><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: right;">1</td><td style = "text-align: left;">2022-09-14T12:44:49</td><td style = "text-align: right;">31.7219</td><td style = "text-align: right;">41.7497</td><td style = "text-align: right;">8.85311</td><td style = "text-align: right;">6.227</td><td style = "text-align: right;">0.607172</td><td style = "text-align: right;">28.5445</td><td style = "text-align: right;">66.4384</td><td style = "text-align: right;">-173.71</td><td style = "text-align: right;">1.20753e6</td><td style = "text-align: right;">-1.68165e6</td><td style = "text-align: right;">0.0</td><td style = "text-align: right;">0.0102782</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: right;">1</td><td style = "text-align: left;">2022-09-14T13:59:19</td><td style = "text-align: right;">27.3962</td><td style = "text-align: right;">37.3585</td><td style = "text-align: right;">8.36951</td><td style = "text-align: right;">6.02703</td><td style = "text-align: right;">0.566405</td><td style = "text-align: right;">27.3084</td><td style = "text-align: right;">70.5095</td><td style = "text-align: right;">-153.641</td><td style = "text-align: right;">1.20779e6</td><td style = "text-align: right;">-1.68216e6</td><td style = "text-align: right;">0.0</td><td style = "text-align: right;">0.0104143</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">3</td><td style = "text-align: right;">1</td><td style = "text-align: left;">2022-09-15T12:44:49</td><td style = "text-align: right;">31.7219</td><td style = "text-align: right;">41.7497</td><td style = "text-align: right;">8.48116</td><td style = "text-align: right;">6.45538</td><td style = "text-align: right;">0.479664</td><td style = "text-align: right;">28.7566</td><td style = "text-align: right;">71.483</td><td style = "text-align: right;">-131.167</td><td style = "text-align: right;">1.20574e6</td><td style = "text-align: right;">-1.68088e6</td><td style = "font-style: italic; text-align: right;">missing</td><td style = "font-style: italic; text-align: right;">missing</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">4</td><td style = "text-align: right;">2</td><td style = "text-align: left;">2022-09-14T12:44:49</td><td style = "text-align: right;">84.6792</td><td style = "text-align: right;">101.523</td><td style = "text-align: right;">20.8234</td><td style = "text-align: right;">5.61064</td><td style = "text-align: right;">-0.316888</td><td style = "text-align: right;">48.9224</td><td style = "text-align: right;">66.5149</td><td style = "text-align: right;">-173.725</td><td style = "text-align: right;">1.19448e6</td><td style = "text-align: right;">-1.67628e6</td><td style = "text-align: right;">0.0</td><td style = "text-align: right;">0.00823861</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">5</td><td style = "text-align: right;">2</td><td style = "text-align: left;">2022-09-14T13:59:19</td><td style = "text-align: right;">77.0764</td><td style = "text-align: right;">94.5103</td><td style = "text-align: right;">20.5266</td><td style = "text-align: right;">5.27752</td><td style = "text-align: right;">-0.340923</td><td style = "text-align: right;">47.8983</td><td style = "text-align: right;">71.4251</td><td style = "text-align: right;">-127.612</td><td style = "text-align: right;">1.19499e6</td><td style = "text-align: right;">-1.67704e6</td><td style = "text-align: right;">0.0</td><td style = "text-align: right;">0.0126095</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">6</td><td style = "text-align: right;">2</td><td style = "text-align: left;">2022-09-15T12:44:49</td><td style = "text-align: right;">84.6136</td><td style = "text-align: right;">102.441</td><td style = "text-align: right;">21.0872</td><td style = "text-align: right;">5.65186</td><td style = "text-align: right;">-0.358114</td><td style = "text-align: right;">49.4344</td><td style = "text-align: right;">71.5446</td><td style = "text-align: right;">-131.022</td><td style = "text-align: right;">1.19268e6</td><td style = "text-align: right;">-1.67576e6</td><td style = "font-style: italic; text-align: right;">missing</td><td style = "font-style: italic; text-align: right;">missing</td></tr></tbody></table></div>

The other 'goodness of fit' metrics are being excluded from the output as they can be easily computed from the inputs. I was thinking that the same is true for centroid distances (thus not included above). Should they? Any further thoughts? Thanks!

**Comment by danielmwatkins** [comment]
Yes, along-track distance is easily calculated from the other columns. I
think Float64 makes sense for area_under and corr as well, since we want to
be able to have NaN for the last timestamp in a trajectory. For clarity, I
wonder if we should make the names "area_under" and "corr" more
descriptive. The column as listed is actual 1 - correlation, is it not? If
so, we should convert it to correlation (so the first row would be ~0.99
instead of 0.01) since in the Lopez-Acosta paper psi-s correlation is
described with large values being a better match. For area_under, that's
the total area of mismatch isn't it? Like, a 0 means that after rotation,
the floes line up perfectly? If so, instead of calling it "area_under" we
could call it "area_mismatch". Thoughts?

**Comment by cpaniaguam** [comment]
> The column as listed is actual 1 - correlation, is it not? If so, we should convert it to correlation (so the first row would be ~0.99 instead of 0.01) since in the Lopez-Acosta paper psi-s correlation is described with large values being a better match. For area_under, that's the total area of mismatch isn't it? Like, a 0 means that after rotation, the floes line up perfectly? If so, instead of calling it "area_under" we could call it "area_mismatch". Thoughts?

@danielmwatkins 
I think those are very sensible choices; I will implement these changes. I now have some real test data I think we could use to thoroughly test the tracker and these new features. Thanks again!

File: src/tracker/tracker.jl [hunk 1]
```diff
@@ -1,41 +1,58 @@
...
+    end
+end
+
+function _pairfloes(
     segmented_imgs::Vector{BitMatrix},
     props::Vector{DataFrame},
     passtimes::Vector{DateTime},
-    dt::Vector{Float64},
     condition_thresholds,
     mc_thresholds,
```
Comment (tdivoll) [inline src/tracker/tracker.jl:52]:
```suggestion
    mc_thresholds
```

File: src/tracker/tracker.jl [hunk 2]
```diff
@@ -118,72 +135,111 @@ function pairfloes(
...
+    rename!(rightdf, Dict(zip(rightcolnames, leftcolnames)))
 
-    # Merge all props into one long DataFrame
-    propsvert = vcat(props...)
+    _pairs = vcat(leftdf, rightdf)
 
-    # rename uuid to ID
-    rename!(propsvert, :uuid => :ID)
+    # sort by uuid_0, passtime and keep unique rows
+    _pairs = DataFrames.sort!(_pairs, [:uuid_0, :passtime]) |> unique
```
Comment (tdivoll) [inline src/tracker/tracker.jl:168]: Nice! That's a great way to just get rid of the extra rows

File: src/tracker/tracker-funcs.jl [hunk 3]
```diff
@@ -497,3 +497,58 @@ function addfloemasks!(props, imgs)
     end
     return nothing
 end
+
+## LatLon functions originally from IFTPipeline.jl
```
Comment (tdivoll) [inline src/tracker/tracker-funcs.jl:501]: Will it be removed from `ice-floe-tracker-pipeline`? It looks like these were orginially in the `h5.jl` script in that repo.

**Review by tdivoll — CHANGES_REQUESTED** [review]
Looks good! Maybe just update the comment about where these functions were originally and log an issue for removing from the other repo.

File: src/tracker/tracker-funcs.jl (hunk 3 above)
Comment (cpaniaguam) [reply src/tracker/tracker-funcs.jl:501]: https://github.com/WilhelmusLab/ice-floe-tracker-pipeline/pull/55

Comment (cpaniaguam) [reply src/tracker/tracker-funcs.jl:501]: https://github.com/WilhelmusLab/ice-floe-tracker-pipeline/issues/90

**Review by tdivoll — APPROVED** [review]
Looks good to me!

**Pull request merged**